src/NServiceBus.Core/OpenTelemetry/Metrics/MetricsExtensions.cs

```diff
@@ -0,0 +1,48 @@
+namespace NServiceBus;
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+static class MetricsExtensions
+{
+    public static bool TryGetTimeSent(this ReceivePipelineCompleted completed, out DateTimeOffset timeSent)
+    {
+        var headers = completed.ProcessedMessage.Headers;
+        if (headers.TryGetValue(Headers.TimeSent, out var timeSentString))
+        {
+            timeSent = DateTimeOffsetHelper.ToDateTimeOffset(timeSentString);
+            return true;
+        }
+        timeSent = DateTimeOffset.MinValue;
+        return false;
+    }
+
+    public static bool TryGetDeliverAt(this ReceivePipelineCompleted completed, out DateTimeOffset deliverAt)
+    {
+        var headers = completed.ProcessedMessage.Headers;
+        if (headers.TryGetValue(Headers.DeliverAt, out var deliverAtString))
+        {
+            deliverAt = DateTimeOffsetHelper.ToDateTimeOffset(deliverAtString);
+            return true;
+        }
+        deliverAt = DateTimeOffset.MinValue;
+        return false;
+    }
+
+    public static bool TryGetMessageType(this ReceivePipelineCompleted completed, out string processedMessageType)
+        => completed.ProcessedMessage.Headers.TryGetMessageType(out processedMessageType);
+    static bool TryGetMessageType(this Dictionary<string, string> headers, out string processedMessageType)
+    {
+        if (headers.TryGetValue(Headers.EnclosedMessageTypes, out var enclosedMessageType))
+        {
+            string messageTypeHeader = !string.IsNullOrEmpty(enclosedMessageType) ? enclosedMessageType.Split(';').FirstOrDefault() : default;
+            string messageTypeName = !string.IsNullOrEmpty(messageTypeHeader) ? messageTypeHeader.Split(',').FirstOrDefault() : default;
```

Comment:
I think the Span variant of this would look something like…

```csharp
var enclosedTypesSpan = enclosedMessageType.AsSpan();
var firstComma = enclosedTypesSpan.IndexOf(',');
var messageType = enclosedTypesSpan.Slice(0, firstComma).ToString();
```

But have to be careful. It could be possible to have a native message sent with an enclosed message types that does not contain the `;` separator (no `ICommand` or `IEvent` either) and it would also be possible to have a native message have either `My.Namespace.UserType, MyAssembly` or even just `My.Namespace.UserType`.

Reply:
[This PR](https://github.com/Particular/NServiceBus/pull/7064)  could prevent the need to reconstruct data from headers, having the opportunity to automatically get from the Metrics Bag the Message Type reported by the InvokeHandlerTerminator.

Reply:
I think that this will be addressed by https://github.com/Particular/NServiceBus/pull/7077

Reply:
Ok, I guess I don't understand in what order the PRs are going to be merged, but I trust we won't be leaving String.Split() stuff in the pipeline hot path.

Reply:
I converted all the metrics PR to a draft, waiting to merge #7077. Once that's merged, I'll rebase all three metrics PR.